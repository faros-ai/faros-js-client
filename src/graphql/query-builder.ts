import {EnumType, jsonToGraphQLQuery} from 'json-to-graphql-query';
import {isNil} from 'lodash';

import {
  ConflictClause,
  DeleteMutationObject,
  Mutation,
  UpsertMutationObject,
} from './types';

type MutationFieldValue =
  | string
  | number
  | boolean
  | any[]
  | {category: string; detail: string}
  | {[field: string]: string | number | boolean}
  | Ref
  | undefined
  | null;

interface MutationFields {
  [field: string]: MutationFieldValue;
}

interface Conflict {
  constraint: string;
  update_columns: string[];
}

export interface FarosModel {
  [modelName: string]: MutationFields;
}

export class Ref {
  constructor(readonly model?: FarosModel) {}
}

export class QueryBuilder {
  constructor(private readonly origin: string) {}

  /**
   * Creates an upsert mutation for the provided Faros model.
   * @param model   The Faros model
   * @param conflict Override the default conflict clause
   * @returns       The upsert mutation
   */
  upsert(model: FarosModel, conflict?: Conflict): Mutation {
    const mutationObj = this.upsertMutationObj({model, ref: false, conflict});
    const modelName = Object.keys(model)[0];
    return {
      mutation: {
        [`insert_${modelName}_one`]: {__args: mutationObj, id: true},
      },
    };
  }

  delete(model: FarosModel): Mutation {
    const deleteObj = this.deleteMutationObj(model);
    const modelName = Object.keys(model)[0];
    return {
      mutation: {
        [`delete_${modelName}`]: {
          __args: deleteObj,
          returning: {
            id: true,
          },
        },
      },
    };
  }

  /**
   * Creates a Ref that can be used in another Faros Model.
   */
  ref(model?: FarosModel): Ref {
    return new Ref(model);
  }

  /**
   * Creates a mutation object that will update every field of the Faros Model.
   * If the model contains fields that reference another
   * model, those fields will be recursively turned into MutationReferences.
   * @param model   The Faros model
   * @param ref     If the mutationObj should be a reference
   * @returns       The mutation object
   */
  private upsertMutationObj(args: {
    model: FarosModel;
    conflict?: Conflict;
    ref?: boolean;
  }): UpsertMutationObject {
    const {model, conflict} = args;
    const ref = args.ref ?? false;

    const [modelName, fields] = Object.entries(model)[0];
    const mutObj: any = {};
    const mask = ['refreshedAt'];

    for (const [k, v] of Object.entries(fields ?? {})) {
      let maskKey = k;
      if (isNil(v)) {
        mutObj[k] = null;
      } else if (v instanceof Ref) {
        if (v.model) {
          mutObj[k] = this.upsertMutationObj({model: v.model, ref: true});
        } else {
          mutObj[k] = null;
        }
        // ref's key should be suffixed with Id for onConflict field
        maskKey += 'Id';
      } else {
        mutObj[k] = Array.isArray(v) ? arrayLiteral(v) : v;
      }
      if (!ref) {
        mask.push(maskKey);
      }
    }

    if (!ref) {
      mutObj.origin = this.origin;
      conflict?.update_columns?.push('origin');
      mask.push('origin');
    }

    const defaultConflict = {
      constraint: `${modelName}_pkey`,
      update_columns: mask,
    };

    return {
      [!ref ? 'object' : 'data']: mutObj,
      on_conflict: this.createConflictClause(conflict ?? defaultConflict),
    };
  }

  /**
   * Creates a delete mutation object that will create _eq keys for every field
   * for the Faros Model.
   *
   * @param model   The Faros model
   * @param ref     If the DeleteMutationObject is a reference
   * @returns       The mutation object
   */
  private deleteMutationObj(
    model: FarosModel,
    ref = false
  ): DeleteMutationObject {
    const fields = Object.values(model)[0];
    const mutObj: any = {};

    for (const [k, v] of Object.entries(fields ?? {})) {
      if (isNil(v)) {
        mutObj[k] = null;
      } else if (v instanceof Ref) {
        if (v.model) {
          mutObj[k] = this.deleteMutationObj(v.model, true);
        } else {
          mutObj[k + 'Id'] = {_is_null: true};
        }
      } else {
        const value = Array.isArray(v) ? arrayLiteral(v) : v;
        mutObj[k] = {_eq: value};
      }
    }

    return !ref ? {where: mutObj} : mutObj;
  }

  private createConflictClause(conflict: Conflict): ConflictClause {
    return {
      constraint: new EnumType(conflict.constraint),
      update_columns: conflict.update_columns.map((c) => new EnumType(c)),
    };
  }
}

/**
 * Convert string arrays into postgres literal array.
 * Leave non-string arrays untouched.
 */
export function arrayLiteral(arr: any[]): string | object {
  for (const item of arr) {
    if (typeof item !== 'string') {
      return arr;
    }
  }
  return `{${arr
    .map((s) => {
      const pre = s.startsWith('"') ? '' : '"';
      const suf = s.endsWith('"') ? '' : '"';
      return `${pre}${s}${suf}`;
    })
    .join(',')}}`;
}

export function mask(object: any): string[] {
  return Object.keys(object);
}

/**
 * Constructs a gql query from an array of json mutations.
 * The outputted qql mutation might look like:
 *
 *   mutation  {
 *     i1: insert_cicd_Artifact_one(object: {uid: "u1b"}) {
 *       id
 *       refreshedAt
 *     }
 *     i2: insert_cicd_Artifact_one(object: {uid: "u2b"}) {
 *       id
 *       refreshedAt
 *     }
 *   }
 *
 *  Notable here are the i1/i2 aliases.
 *  These are required when multiple operations share the same
 *  name (e.g. insert_cicd_Artifact_one) and are supported in
 *  jsonToGraphQLQuery with __aliasFor directive.
 *
 *  @return batch gql mutation or undefined if the input is undefined, empty
 *  or doesn't contain any mutations.
 */
export function batchMutation(mutations: Mutation[]): string | undefined {
  if (mutations.length) {
    const queryObj: any = {};
    mutations.forEach((query, idx) => {
      if (query.mutation) {
        const queryType = Object.keys(query.mutation)[0];
        const queryBody = query.mutation[queryType];
        queryObj[`m${idx}`] = {
          __aliasFor: queryType,
          ...queryBody,
        };
      }
    });
    if (Object.keys(queryObj).length > 0) {
      return jsonToGraphQLQuery({mutation: queryObj});
    }
  }
  return undefined;
}
