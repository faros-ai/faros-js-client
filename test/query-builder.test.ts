import * as sut from '../src/graphql/query-builder';

const ORIGIN = 'test-origin';

describe('query builder', () => {
  const qb = new sut.QueryBuilder(ORIGIN);

  const compute_Application = {
    name: '<application_name>',
    platform: '<application_platform>',
  };
  const cicd_Organization = {
    uid: '<organization_uid>',
    source: '<organization_source>',
  };
  const cicd_Pipeline = {
    uid: '<pipeline_uid>',
    organization: qb.ref({cicd_Organization}),
  };
  const cicd_Build = {
    uid: '<cicd_Build>',
    pipeline: qb.ref({cicd_Pipeline}),
    name: '<build_name>',
  };
  const cicd_Deployment = {
    uid: '<deployment_uid',
    source: '<deployment_source>',
    application: qb.ref({compute_Application}),
    build: qb.ref({cicd_Build}),
    status: {
      category: 'Success',
      detail: '<status_detail>',
    },
  };

  test('creates mutations', () => {
    const mutations = [
      qb.upsert({compute_Application}),
      qb.upsert({cicd_Organization}),
      qb.upsert({cicd_Pipeline}),
      qb.upsert({cicd_Build}),
      qb.upsert({cicd_Deployment}),
    ];
    const queryString = sut.batchMutation(mutations);
    expect(queryString).toMatchSnapshot();
  });

  test('creates mutations with non-model objects', () => {
    const qa_TestCase = {
      uid: '<uid>',
      source: '<source>',
      name: '<name>',
      before: [{description: '<description>', condition: '<condition>'}],
      after: [{description: '<description>', condition: '<condition>'}],
      tags: ['tag1', 'tag2'],
      qa_DeviceInfo: {
        name: 'name',
        os: 'os',
        browser: 'browser',
        isSupported: true,
        size: 10,
      },
    };

    const mutations = [qb.upsert({qa_TestCase})];
    const queryString = sut.batchMutation(mutations);
    expect(queryString).toMatchSnapshot();
  });

  test('creates mutations with undefined and null fields', () => {
    const qa_TestCase = {
      uid: '<uid>',
      source: '<source>',
      name: '<name>',
      before: null,
      after: undefined,
      tags: ['tag1', 'tag2'],
    };

    const mutations = [qb.upsert({qa_TestCase})];
    const queryString = sut.batchMutation(mutations);
    expect(queryString).toMatchSnapshot();
  });

  test('delete mutations with non-model objects', () => {
    const qa_TestCase = {
      uid: '<uid>',
      source: '<source>',
      name: '<name>',
      before: null,
      after: undefined,
      tags: ['tag1', 'tag2'],
    };

    const mutations = [qb.delete({qa_TestCase})];
    const queryString = sut.batchMutation(mutations);
    expect(queryString).toMatchSnapshot();
  });

  test('delete mutations with model refs', () => {
    const mutations = [qb.delete({cicd_Deployment}), qb.delete({cicd_Build})];
    const queryString = sut.batchMutation(mutations);
    expect(queryString).toMatchSnapshot();
  });

  test('upsert undefined ref', () => {
    const cicd_Pipeline = {
      uid: '<pipeline_uid>',
      organization: qb.ref(undefined),
    };
    const mutations = [qb.upsert({cicd_Pipeline})];
    const queryString = sut.batchMutation(mutations);
    expect(queryString).toMatchSnapshot();
  });

  test('delete undefined ref', () => {
    const cicd_Pipeline = {
      uid: '<pipeline_uid>',
      organization: qb.ref(undefined),
    };
    const mutations = [qb.delete({cicd_Pipeline})];
    const queryString = sut.batchMutation(mutations);
    expect(queryString).toMatchSnapshot();
  });

  test('upsert with conflict override', () => {
    const org_ApplicationOwnership = {
      team: qb.ref({org_Team: {uid: 'test_team'}}),
      application: qb.ref({
        compute_Application: {name: 'test_app', platform: 'test_platform'},
      }),
    };
    const mutations = [
      qb.upsert(
        {org_ApplicationOwnership},
        {
          constraint: 'org_ApplicationOwnership_application_id_unique',
          update_columns: ['teamId'],
        }
      ),
    ];
    const queryString = sut.batchMutation(mutations);
    expect(queryString).toMatchSnapshot();
  });
});

describe('arrayLiteral', () => {
  test('strings', () => {
    expect(sut.arrayLiteral(['a', 'b', 'c'])).toEqual('{"a","b","c"}');
    expect(sut.arrayLiteral(['"a', 'b"', '"c"'])).toEqual('{"a","b","c"}');
  });
});
