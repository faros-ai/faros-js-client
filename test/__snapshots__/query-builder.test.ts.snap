// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query builder creates mutations 1`] = `"mutation { m0: insert_compute_Application_one (object: {name: "<application_name>", platform: "<application_platform>", origin: "test-origin"}, on_conflict: {constraint: compute_Application_pkey, update_columns: [refreshedAt, name, platform, origin]}) { id } m1: insert_cicd_Organization_one (object: {uid: "<organization_uid>", source: "<organization_source>", origin: "test-origin"}, on_conflict: {constraint: cicd_Organization_pkey, update_columns: [refreshedAt, uid, source, origin]}) { id } m2: insert_cicd_Pipeline_one (object: {uid: "<pipeline_uid>", organization: {data: {uid: "<organization_uid>", source: "<organization_source>"}, on_conflict: {constraint: cicd_Organization_pkey, update_columns: [refreshedAt]}}, origin: "test-origin"}, on_conflict: {constraint: cicd_Pipeline_pkey, update_columns: [refreshedAt, uid, organizationId, origin]}) { id } m3: insert_cicd_Build_one (object: {uid: "<cicd_Build>", pipeline: {data: {uid: "<pipeline_uid>", organization: {data: {uid: "<organization_uid>", source: "<organization_source>"}, on_conflict: {constraint: cicd_Organization_pkey, update_columns: [refreshedAt]}}}, on_conflict: {constraint: cicd_Pipeline_pkey, update_columns: [refreshedAt]}}, name: "<build_name>", origin: "test-origin"}, on_conflict: {constraint: cicd_Build_pkey, update_columns: [refreshedAt, uid, pipelineId, name, origin]}) { id } m4: insert_cicd_Deployment_one (object: {uid: "<deployment_uid", source: "<deployment_source>", application: {data: {name: "<application_name>", platform: "<application_platform>"}, on_conflict: {constraint: compute_Application_pkey, update_columns: [refreshedAt]}}, build: {data: {uid: "<cicd_Build>", pipeline: {data: {uid: "<pipeline_uid>", organization: {data: {uid: "<organization_uid>", source: "<organization_source>"}, on_conflict: {constraint: cicd_Organization_pkey, update_columns: [refreshedAt]}}}, on_conflict: {constraint: cicd_Pipeline_pkey, update_columns: [refreshedAt]}}, name: "<build_name>"}, on_conflict: {constraint: cicd_Build_pkey, update_columns: [refreshedAt]}}, status: {category: "Success", detail: "<status_detail>"}, origin: "test-origin"}, on_conflict: {constraint: cicd_Deployment_pkey, update_columns: [refreshedAt, uid, source, applicationId, buildId, status, origin]}) { id } }"`;

exports[`query builder creates mutations with non-model objects 1`] = `"mutation { m0: insert_qa_TestCase_one (object: {uid: "<uid>", source: "<source>", name: "<name>", before: [{description: "<description>", condition: "<condition>"}], after: [{description: "<description>", condition: "<condition>"}], tags: "{\\"tag1\\",\\"tag2\\"}", qa_DeviceInfo: {name: "name", os: "os", browser: "browser", isSupported: true, size: 10}, origin: "test-origin"}, on_conflict: {constraint: qa_TestCase_pkey, update_columns: [refreshedAt, uid, source, name, before, after, tags, qa_DeviceInfo, origin]}) { id } }"`;

exports[`query builder creates mutations with undefined and null fields 1`] = `"mutation { m0: insert_qa_TestCase_one (object: {uid: "<uid>", source: "<source>", name: "<name>", before: null, after: null, tags: "{\\"tag1\\",\\"tag2\\"}", origin: "test-origin"}, on_conflict: {constraint: qa_TestCase_pkey, update_columns: [refreshedAt, uid, source, name, before, after, tags, origin]}) { id } }"`;

exports[`query builder delete mutations with model refs 1`] = `"mutation { m0: delete_cicd_Deployment (where: {uid: {_eq: "<deployment_uid"}, source: {_eq: "<deployment_source>"}, application: {name: {_eq: "<application_name>"}, platform: {_eq: "<application_platform>"}}, build: {uid: {_eq: "<cicd_Build>"}, pipeline: {uid: {_eq: "<pipeline_uid>"}, organization: {uid: {_eq: "<organization_uid>"}, source: {_eq: "<organization_source>"}}}, name: {_eq: "<build_name>"}}, status: {_eq: {category: "Success", detail: "<status_detail>"}}}) { returning { id } } m1: delete_cicd_Build (where: {uid: {_eq: "<cicd_Build>"}, pipeline: {uid: {_eq: "<pipeline_uid>"}, organization: {uid: {_eq: "<organization_uid>"}, source: {_eq: "<organization_source>"}}}, name: {_eq: "<build_name>"}}) { returning { id } } }"`;

exports[`query builder delete mutations with non-model objects 1`] = `"mutation { m0: delete_qa_TestCase (where: {uid: {_eq: "<uid>"}, source: {_eq: "<source>"}, name: {_eq: "<name>"}, before: null, after: null, tags: {_eq: "{\\"tag1\\",\\"tag2\\"}"}}) { returning { id } } }"`;

exports[`query builder delete undefined ref 1`] = `"mutation { m0: delete_cicd_Pipeline (where: {uid: {_eq: "<pipeline_uid>"}, organizationId: {_is_null: true}}) { returning { id } } }"`;

exports[`query builder upsert undefined ref 1`] = `"mutation { m0: insert_cicd_Pipeline_one (object: {uid: "<pipeline_uid>", organization: null, origin: "test-origin"}, on_conflict: {constraint: cicd_Pipeline_pkey, update_columns: [refreshedAt, uid, organizationId, origin]}) { id } }"`;

exports[`query builder upsert with conflict override 1`] = `"mutation { m0: insert_org_ApplicationOwnership_one (object: {team: {data: {uid: "test_team"}, on_conflict: {constraint: org_Team_pkey, update_columns: [refreshedAt]}}, application: {data: {name: "test_app", platform: "test_platform"}, on_conflict: {constraint: compute_Application_pkey, update_columns: [refreshedAt]}}, origin: "test-origin"}, on_conflict: {constraint: org_ApplicationOwnership_application_id_unique, update_columns: [teamId, refreshedAt, origin]}) { id } }"`;
